src/coreclr/inc/gcinfotypes.h

```diff
@@ -612,274 +617,313 @@ void FASTCALL decodeCallPattern(int         pattern,
 #ifndef TARGET_POINTER_SIZE
 #define TARGET_POINTER_SIZE 8    // equal to sizeof(void*) and the managed pointer size in bytes for this target
 #endif
-#define NUM_NORM_CODE_OFFSETS_PER_CHUNK (64)
-#define NUM_NORM_CODE_OFFSETS_PER_CHUNK_LOG2 (6)
-#define NORMALIZE_STACK_SLOT(x) ((x)>>3)
-#define DENORMALIZE_STACK_SLOT(x) ((x)<<3)
-#define NORMALIZE_CODE_LENGTH(x) (x)
-#define DENORMALIZE_CODE_LENGTH(x) (x)
-// Encode RBP as 0
-#define NORMALIZE_STACK_BASE_REGISTER(x) ((x) ^ 5)
-#define DENORMALIZE_STACK_BASE_REGISTER(x) ((x) ^ 5)
-#define NORMALIZE_SIZE_OF_STACK_AREA(x) ((x)>>3)
-#define DENORMALIZE_SIZE_OF_STACK_AREA(x) ((x)<<3)
-#define CODE_OFFSETS_NEED_NORMALIZATION 0
-#define NORMALIZE_CODE_OFFSET(x) (x)
-#define DENORMALIZE_CODE_OFFSET(x) (x)
-#define NORMALIZE_REGISTER(x) (x)
-#define DENORMALIZE_REGISTER(x) (x)
-#define NORMALIZE_NUM_SAFE_POINTS(x) (x)
-#define DENORMALIZE_NUM_SAFE_POINTS(x) (x)
-#define NORMALIZE_NUM_INTERRUPTIBLE_RANGES(x) (x)
-#define DENORMALIZE_NUM_INTERRUPTIBLE_RANGES(x) (x)
-
-#define PSP_SYM_STACK_SLOT_ENCBASE 6
-#define GENERICS_INST_CONTEXT_STACK_SLOT_ENCBASE 6
-#define SECURITY_OBJECT_STACK_SLOT_ENCBASE 6
-#define GS_COOKIE_STACK_SLOT_ENCBASE 6
-#define CODE_LENGTH_ENCBASE 8
-#define STACK_BASE_REGISTER_ENCBASE 3
-#define SIZE_OF_STACK_AREA_ENCBASE 3
-#define SIZE_OF_EDIT_AND_CONTINUE_PRESERVED_AREA_ENCBASE 4
-#define REVERSE_PINVOKE_FRAME_ENCBASE 6
-#define NUM_REGISTERS_ENCBASE 2
-#define NUM_STACK_SLOTS_ENCBASE 2
-#define NUM_UNTRACKED_SLOTS_ENCBASE 1
-#define NORM_PROLOG_SIZE_ENCBASE 5
-#define NORM_EPILOG_SIZE_ENCBASE 3
-#define NORM_CODE_OFFSET_DELTA_ENCBASE 3
-#define INTERRUPTIBLE_RANGE_DELTA1_ENCBASE 6
-#define INTERRUPTIBLE_RANGE_DELTA2_ENCBASE 6
-#define REGISTER_ENCBASE 3
-#define REGISTER_DELTA_ENCBASE 2
-#define STACK_SLOT_ENCBASE 6
-#define STACK_SLOT_DELTA_ENCBASE 4
-#define NUM_SAFE_POINTS_ENCBASE 2
-#define NUM_INTERRUPTIBLE_RANGES_ENCBASE 1
-#define NUM_EH_CLAUSES_ENCBASE 2
-#define POINTER_SIZE_ENCBASE 3
-#define LIVESTATE_RLE_RUN_ENCBASE 2
-#define LIVESTATE_RLE_SKIP_ENCBASE 4
+
+#define TargetGcInfoEncoding AMD64GcInfoEncoding
+
+struct AMD64GcInfoEncoding {
+    static const uint32_t NUM_NORM_CODE_OFFSETS_PER_CHUNK = (64);
+
+    static const uint32_t NUM_NORM_CODE_OFFSETS_PER_CHUNK_LOG2 = (6);
+    template <typename T> static inline constexpr T NORMALIZE_STACK_SLOT (T x) { return ((x)>>3); }
```

Comment:
Can these take and return concrete type like `uint32_t` instead of being templates?

Reply:
Yes, but it requires intrusive changes to their callers, so I decided to revert it.

Reply:
Specifically https://github.com/dotnet/runtime/pull/113223/commits/bf03fed72a1e1bdfb9bc1f5d19126afe34e5eb65

Reply:
> Specifically https://github.com/dotnet/runtime/commit/bf03fed72a1e1bdfb9bc1f5d19126afe34e5eb65

Are these all changes needed to make it work? I do not see a problem with it - it does not look like a very intrusive change.

Reply:
They were mostly to resolve build failures on windows and linux, mostly msvc. If you're fine with those changes to the gc stuff I'll go with that approach instead of the templates.

I wasn't sure how robust our test coverage is for this code